src/base/lockfree_freelist.hpp

```diff
@@ -0,0 +1,392 @@
+  }
+
+  template <class T>
+  void
+  freelist<T>::alloc_backbuffer ()
+  {
+    T *new_bb_head = NULL;
+    T *new_bb_tail = NULL;
+
+    alloc_list (new_bb_head, new_bb_tail);
+
+    // update backbuffer tail
+    T *dummy_null = NULL;
+    m_backbuffer_tail.compare_exchange_strong (dummy_null, new_bb_tail);
+
+    // update backbuffer head
+    push_to_list (new_bb_head, new_bb_tail, m_backbuffer_head);
+    m_bb_count += m_block_size;
+  }
+
+  template <class T>
+  void
+  freelist<T>::force_alloc_block ()
+  {
+    T *new_head = NULL;
+    T *new_tail = NULL;
+    alloc_list (new_head, new_tail);
+
+    // push directly to available
+    push_to_list (new_head, new_tail, m_available_list);
+    m_available_count += m_block_size;
+    ++m_forced_alloc_count;
+  }
+
+  template <class T>
+  void
+  freelist<T>::alloc_list (T *&head, T *&tail)
+  {
+    head = tail = NULL;
+    T *t;
+    for (size_t i = 0; i < m_block_size; i++)
+      {
+	t = new T ();
+	if (tail == NULL)
+	  {
+	    tail = t;
+	  }
+	t->get_freelist_link ().store (head);
+	head = t;
+      }
+    m_alloc_count += m_block_size;
+  }
+
+  template <class T>
+  void
+  freelist<T>::dealloc_list (T *head)
+  {
+    // free all
+    T *save_next = NULL;
+    for (T *t = head; t != NULL; t = save_next)
+      {
+	save_next = t->get_freelist_link ().load ();
+	delete t;
+      }
+  }
+
+  template <class T>
+  freelist<T>::~freelist ()
+  {
+    clear ();
+  }
+
+  template <class T>
+  void
+  freelist<T>::clear ()
+  {
+    final_sanity_checks ();
+
+    // move back-buffer to available
+    dealloc_list (m_backbuffer_head.load ());
+    m_backbuffer_head = NULL;
+    m_backbuffer_tail = NULL;
+
+    dealloc_list (m_available_list.load ());
+    m_available_list = NULL;
+
+    m_available_count = m_bb_count = m_alloc_count = 0;
+  }
+
+  template <class T>
+  T *
+  freelist<T>::claim ()
+  {
+    // todo: make sure transaction is open here
+    T *t;
+    size_t count = 0;
+    for (t = pop_from_available (); t == NULL && count < 100; t = pop_from_available (), ++count)
+      {
+	// if it loops many times, it is probably because the back-buffer allocator was preempted for a very long time.
+	// force allocations
+	swap_backbuffer ();
+      }
+    // if swapping backbuffer didn't work (probably back-buffer allocator was preempted for a long time), force
+    // allocating directly into available list
+    while (t == NULL)
+      {
+	force_alloc_block ();
+	t = pop_from_available ();
+      }
+    assert (t != NULL);
+    m_available_count--;
+    return t;
+  }
+
+  template <class T>
+  T *
+  freelist<T>::pop_from_available ()
+  {
+    T *rhead = NULL;
+    T *rhead_copy = NULL;
+    T *next;
+    do
+      {
+	rhead = m_available_list;
+	if (rhead == NULL)
+	  {
+	    return NULL;
+	  }
+	next = rhead->get_freelist_link ().load ();
+	rhead_copy = rhead;
+	// todo: this is a dangerous preemption point; if I am preempted here, and thread 2 comes and does:
```

Comment:
Is it possible to avoid this by transforming available_list from a stack to queue:  having m_available_list_head and m_available_list_tail and change 'claim' to pop from head and 'retire' to push to tail ?

Reply:
I am not sure. I don't know how to safely manage two pointers - head and tail - with these operations. Claim/retire may have to update both values in corner cases.

The existing implementation relies on having to change a single pointer.